### PR TITLE
Amazon Q: Added telemetry metrics and modified customer facing messages for Q Codescans.

### PR DESCRIPTION
**Pull request by laileni-aws**
## Problem
- Inconsistent customer facing messages and telemetry events in codescan.

## Solution
- Added customer facing error messages by removing existing to keep consistent.
>- `Amazon Q: You have reached the monthly limit for project scans.`
>- `Amazon Q encountered an error while scanning for security issues. Please try again later.`
>- `Amazon Q: No workspace folders found`
>- `Amazon Q: Project does not contain valid files to scan`
>- `Amazon Q: The selected project exceeds the input artifact limit. Try again with a smaller project. For more information about scan limits, see the [Amazon Q documentation](https://docs.aws.amazon.com/amazonq/latest/qdeveloper-ug/security-scans.html#quotas).`
>- `Amazon Q: The selected file exceeds the input artifact limit. Try again with a smaller file. For more information about scan limits, see the [Amazon Q documentation](https://docs.aws.amazon.com/amazonq/latest/qdeveloper-ug/security-scans.html#quotas).`
>- `Amazon Q is unable to upload your workspace artifacts to Amazon S3 for security scans. For more information, see the [Amazon Q documentation](https://docs.aws.amazon.com/amazonq/latest/qdeveloper-ug/security_iam_manage-access-with-policies.html#data-perimeters).`
- Added necessary telemetry events to understand the customer facing errors at different stages in Kibana.

## License

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: packages/core/src/codewhisperer/service/securityScanHandler.ts [hunk 1]
```diff
@@ -133,8 +133,8 @@ export async function pollScanJobStatus(
                 : CodeWhispererConstants.codeScanJobTimeoutSeconds
         if (timer > timeoutSeconds) {
             logger.verbose(`Scan job status: ${status}`)
-            logger.verbose(`Scan job timeout.`)
-            throw new Error('Scan job timeout.')
+            logger.verbose(`Security Scan failed. Amazon Q timed out.`)
+            throw new Error('Security Scan failed. Amazon Q timed out.')
```
Comment (ctlai95) [inline packages/core/src/codewhisperer/service/securityScanHandler.ts:137]: I think the main catch block already says "Security scan failed.", so this change might cause a duplicate message

File: packages/core/src/codewhisperer/commands/startSecurityScan.ts [hunk 2]
```diff
@@ -172,7 +172,7 @@ export async function startSecurityScan(
             scanName
         )
         if (scanJob.status === 'Failed') {
-            throw new Error(scanJob.errorMessage)
+            throw new Error('Failed code scan service error')
```
[the remark this answers is not in the document]
Comment (laileni-aws) [reply packages/core/src/codewhisperer/commands/startSecurityScan.ts:176]: I will revert this accordingly.

File: packages/toolkit/.changes/next-release/Bug Fix-ebe770da-054d-43e7-8141-cd32b6159e3e.json
Comment (ctlai95) [inline packages/toolkit/.changes/next-release/Bug Fix-ebe770da-054d-43e7-8141-cd32b6159e3e.json:1]: Changelog should go in `amazonq` folder

File: packages/toolkit/.changes/next-release/Bug Fix-ebe770da-054d-43e7-8141-cd32b6159e3e.json [hunk 3]
```diff
@@ -0,0 +1,4 @@
+{
+	"type": "Bug Fix",
+	"description": "Amazon Q: Added telemetry metrics and modified customer facing messages for Q codescans."
```
Comment (ctlai95) [inline packages/toolkit/.changes/next-release/Bug Fix-ebe770da-054d-43e7-8141-cd32b6159e3e.json:3]:
```suggestion
	"description": "Security Scan: Improved error notifications"
```

**Review by ctlai95 — APPROVED** [review]

File: packages/core/src/codewhisperer/models/errors.ts [hunk 4]
```diff
@@ -0,0 +1,117 @@
...
+    UploadArtifactToS3Error = 'UploadArtifactToS3Error',
+    DefaultError = 'DefaultError',
+    InvalidSourceZip = 'InvalidSourceZip',
+    CreateCodeScanError = 'CreateCodeScanError',
+    CreateCodeScanFailedError = 'CreateCodeScanFailedError',
+    SecurityScanTimedOutError = 'SecurityScanTimedOutError',
+    CodeScanJobFailedError = 'CodeScanJobFailedError',
+}
+
+export const mapErrorToCustomerFacingMessage: Record<ErrorCodes, string> = {
```
Comment (justinmk3) [inline packages/core/src/codewhisperer/models/errors.ts:30]: why not just store the "customer facing message" on the error? there is way too much over-abstraction and indirection in this PR

File: packages/core/src/codewhisperer/models/errors.ts [hunk 5]
```diff
@@ -0,0 +1,117 @@
...
+
+export class FileSizeExceededError extends ToolkitError {
+    constructor() {
+        super('Payload size limit reached', {
+            code: ErrorCodes.FileSizeExceeded,
+        })
+    }
+}
+
+export class ProjectSizeExceededError extends ToolkitError {
```
Comment (justinmk3) [inline packages/core/src/codewhisperer/models/errors.ts:53]: you are creating custom classes for every error, yet not using the features of a class. the class could store the "customer facing message", and possibly could also avoid the enums and "mapping" indirection

Comment (laileni-aws) [reply packages/core/src/codewhisperer/models/errors.ts]: Created a new `SecurityScanError` class which extends `ToolkitError` to add a new field `customerFacingMessage`. This should work without `ErrorCodes` enum and `mapErrorToCustomerFacingMessage`.

**Review by ctlai95 — APPROVED** [review]

File: packages/core/src/codewhisperer/models/constants.ts [hunk 6]
```diff
@@ -290,9 +290,22 @@ export const freeTierLimitReached = 'You have reached the monthly fair use limit
...
-export const fileScansLimitReached = 'You have reached the monthly quota of auto-scans.'
+export const fileScansLimitReached = 'Amazon Q: You have reached the monthly limit for auto-scans.'
 
-export const projectScansLimitReached = 'You have reached the monthly quota of project scans.'
+export const projectScansLimitReached = 'Amazon Q: You have reached the monthly limit for project scans.'
+
+export const DefaultCodeScanErrorMessage =
+    'Amazon Q encountered an error while scanning for security issues. Try again later.'
+
+export const FileSizeExceededErrorMessage = `Amazon Q: The selected file exceeds the input artifact limit. Try again with a smaller file. For more information about scan limits, see the [Amazon Q documentation](https://docs.aws.amazon.com/amazonq/latest/qdeveloper-ug/security-scans.html#quotas).`
```
Comment (justinmk3) [inline packages/core/src/codewhisperer/models/constants.ts:300]: these could be stored directly on the Error classes.

**Review by justinmk3 — APPROVED** [review]

**Pull request merged**